src/hotspot/share/opto/addnode.cpp

```diff
@@ -1086,166 +1086,98 @@ static bool can_overflow(const TypeInt* t, jint c) {
           (c > 0 && (java_add(t_hi, c) < t_hi)));
 }
 
-// Ideal transformations for MaxINode
-Node* MaxINode::Ideal(PhaseGVN* phase, bool can_reshape) {
-  // Force a right-spline graph
-  Node* l = in(1);
-  Node* r = in(2);
-  // Transform  MaxI1(MaxI2(a, b), c)  into  MaxI1(a, MaxI2(b, c))
-  // to force a right-spline graph for the rest of MaxINode::Ideal().
-  if (l->Opcode() == Op_MaxI) {
-    assert(l != l->in(1), "dead loop in MaxINode::Ideal");
-    r = phase->transform(new MaxINode(l->in(2), r));
-    l = l->in(1);
-    set_req_X(1, l, phase);
-    set_req_X(2, r, phase);
-    return this;
-  }
-
-  // Get left input & constant
-  Node* x = l;
-  jint x_off = 0;
-  if (x->Opcode() == Op_AddI && // Check for "x+c0" and collect constant
-      x->in(2)->is_Con()) {
-    const Type* t = x->in(2)->bottom_type();
-    if (t == Type::TOP) return nullptr;  // No progress
-    x_off = t->is_int()->get_con();
-    x = x->in(1);
-  }
-
-  // Scan a right-spline-tree for MAXs
-  Node* y = r;
-  jint y_off = 0;
-  // Check final part of MAX tree
-  if (y->Opcode() == Op_AddI && // Check for "y+c1" and collect constant
-      y->in(2)->is_Con()) {
-    const Type* t = y->in(2)->bottom_type();
-    if (t == Type::TOP) return nullptr;  // No progress
-    y_off = t->is_int()->get_con();
-    y = y->in(1);
-  }
-  if (x->_idx > y->_idx && r->Opcode() != Op_MaxI) {
-    swap_edges(1, 2);
-    return this;
+// Return:
+// <x, C>,       if n is of the form x + C, where 'C' is a non-TOP constant;
+// <nullptr, _>, if n is of the form x + C, where 'C' is a TOP constant;
+// <n, con>      otherwise.
+static Node* constant_add_input(Node* n, jint* con) {
+  if (n->Opcode() == Op_AddI && n->in(2)->is_Con()) {
+    const Type* t = n->in(2)->bottom_type();
+    if (t == Type::TOP) {
+      return nullptr;
+    }
+    *con = t->is_int()->get_con();
+    n = n->in(1);
   }
+  return n;
+}
```

Comment:
This was confusing to read.

Why not make it a `as_add_constant`, and explicitly always set the `con`:
```
case AddI(x, top): return <nullptr, 0>
case AddI(x, int_con): return <x, int_con>
default: return <n, 0>
```
That would make it easier to argue what the value of `con` is after the call - it does certainly not depend on what it was before.

Reply:
Add a comment that explains that for `top` we will bail out - for that we can check `nullptr`.
In the other cases, we know that `n == AddI(x, int_con)`.

Reply:
Thanks, I went with the tuple return option as suggested and also simplified the semantics of the function. Hope it is clearer now!